### PR TITLE
WIP : Create Identifier Sources API

### PR DESCRIPTION
**Pull request by rehammuzzamil**
Issue # 309
Create Identifier Sources API.

### CODE REVIEW

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+@Controller
+@RequestMapping(value = "/rest/identifiersource")
+public class IdentifierSourceResource {
+
+	private static Logger logger = LoggerFactory.getLogger(IdentifierSourceResource.class.toString());
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:23]: can we use lombok @slf4j

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+	private IdentifierSourceService identifierSourceService;
+
+	protected ObjectMapper objectMapper;
+
+	public static final String IDENTIFIER = "identifier";
+
+	@Autowired
+	public void setObjectMapper(ObjectMapper objectMapper) {
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:32]: any specific reason to use autowiring in different places i.e. setter, constructor etc...

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		this.objectMapper = objectMapper;
+	}
+
+	@Autowired
+	public IdentifierSourceResource(IdentifierSourceService identifierSourceService) {
+		this.identifierSourceService = identifierSourceService;
+	}
+
+	@RequestMapping(value = "getAll", method = RequestMethod.GET, produces = { MediaType.APPLICATION_JSON_VALUE })
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:41]: api url should be consistent... are we using getAll for all other resources? also ideally api should be all lowercase

Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:41]: use Getmapping instead.. we are on spring 5 now..

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	public ResponseEntity<String> get() throws JsonProcessingException {
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:42]: getAll

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		return new ResponseEntity<>(objectMapper.writeValueAsString(
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:43]: can we break it down...

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+				identifierSourceService.findAllIdentifierSources()),
+				RestUtils.getJSONUTF8Headers(), HttpStatus.OK);
+	}
+
+	@RequestMapping(value = "/{identifier}", method = RequestMethod.GET, produces = {
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:48]: use getmapping, postmapping etc..

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+			MediaType.APPLICATION_JSON_VALUE })
+	public ResponseEntity<String> getIdentifierSourceByIdentifier(@PathVariable(IDENTIFIER) String identifier)
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:50]: shorten the name

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+	@RequestMapping(method = RequestMethod.POST, consumes = { MediaType.APPLICATION_JSON_VALUE,
+			MediaType.TEXT_PLAIN_VALUE })
+	public ResponseEntity<String> createOrUpdate(@RequestBody IdentifierSource identifierSource) {
+		try {
+			identifierSourceService.addOrUpdate(identifierSource);
+			return new ResponseEntity<>(HttpStatus.CREATED);
+		}
+		catch (IllegalArgumentException e) {
+			return new ResponseEntity<String>("The request contain illegal argument ", HttpStatus.BAD_REQUEST);
```
Comment (maimoonak) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:67]: where do we validate the idsource... would it throw exception anywhere? could it be handled by GlobalExcHandler... check with Samuel

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java (hunk 1 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:23]: For that, we have to add dependency first

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java (hunk 2 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java]: Updated

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java (hunk 3 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java]: Yes, I found /getAll end-points. I tried to follow the same convention.

Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java]: Done

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java (hunk 6 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java]: Done

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java (hunk 7 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java]: Done

File: src/main/java/org/opensrp/web/config/security/SecurityConfig.java [hunk 9]
```diff
@@ -92,6 +92,7 @@ protected void configure(HttpSecurity http) throws Exception {
 			.mvcMatchers("/rest/viewconfiguration/**").permitAll()
 			.mvcMatchers("/rest/viewconfiguration/**").permitAll()
 			.mvcMatchers("/rest/config/keycloak").permitAll()
+			.mvcMatchers("/rest/identifiersource/**").permitAll()
```
Comment (githengi) [inline src/main/java/org/opensrp/web/config/security/SecurityConfig.java:95]: Remove this, authentication is required to access this

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+				identifierSourceService.findAllIdentifierSources()),
+				RestUtils.getJSONUTF8Headers(), HttpStatus.OK);
+	}
+
+	@GetMapping(value = "/{identifier}", produces = {
+			MediaType.APPLICATION_JSON_VALUE })
+	public ResponseEntity<String> getByIdentifier(@PathVariable String identifier)
+			throws JsonProcessingException {
+
+		return new ResponseEntity<>(objectMapper.writeValueAsString(
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:40]: return the object

File: src/main/java/org/opensrp/web/controller/UniqueIdController.java [hunk 11]
```diff
@@ -123,11 +132,25 @@ boolean checkRoleIfRoleExitst(List<String> roleList, String role) {
...
+				List<String> identifiers = uniqueIdentifierService.generateIdentifiers(identifierSource,
+						Integer.parseInt(numberToGenerate), usedBy);
+				if (identifiers != null) {
+					map.put("identifiers", identifiers);
+				}
+			} else {
+				map.put("identifiers",
+						openmrsIdService.getOpenMRSIdentifiers(source, numberToGenerate, OPENMRS_USER, OPENMRS_PWD));
+			}
+			return new ResponseEntity<>(new Gson().toJson(map), HttpStatus.OK);
```
Comment (githengi) [inline src/main/java/org/opensrp/web/controller/UniqueIdController.java:149]: change to jackson

**Comment by rehammuzzamil** [comment]
https://github.com/OpenSRP/opensrp-server-web/pull/325#discussion_r435247677 
I have removed this authentication, it would be great if you test at your end too to make sure end-point is accessible. @githengi

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+
+@Controller
+@RequestMapping(value = "/rest/identifiersource")
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:21]: change to `/rest/identifier-source`

File: src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+	private static Logger logger = LoggerFactory.getLogger(IdentifierSourceResource.class.toString());
+
+	@Autowired
+	private IdentifierSourceService identifierSourceService;
+
+	@Autowired
+	protected ObjectMapper objectMapper;
+
+	@GetMapping(value = "/getAll", produces = { MediaType.APPLICATION_JSON_VALUE })
```
Comment (githengi) [inline src/main/java/org/opensrp/web/rest/IdentifierSourceResource.java:32]: Map this to the default URL, remove `/getAll`

File: src/main/java/org/opensrp/web/controller/UniqueIdController.java [hunk 14]
```diff
@@ -123,11 +137,25 @@ boolean checkRoleIfRoleExitst(List<String> roleList, String role) {
 		
 		String numberToGenerate = getStringFilter("numberToGenerate", request);
 		String source = getStringFilter("source", request);
+		String usedBy = getStringFilter("usedBy", request);
```
Comment (githengi) [inline src/main/java/org/opensrp/web/controller/UniqueIdController.java:140]: Get this from the authenticated user `authentication.getName()`

File: src/main/java/org/opensrp/web/controller/UniqueIdController.java [hunk 15]
```diff
@@ -123,11 +137,26 @@ boolean checkRoleIfRoleExitst(List<String> roleList, String role) {
 		
 		String numberToGenerate = getStringFilter("numberToGenerate", request);
 		String source = getStringFilter("source", request);
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
```
Comment (githengi) [inline src/main/java/org/opensrp/web/controller/UniqueIdController.java:140]: could you pass authentication as param to this method for easier unit testing of this later

**Pull request merged**